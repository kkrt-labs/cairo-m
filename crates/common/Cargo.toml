[package]
name = "cairo-m-common"
version = "0.1.0"
edition.workspace = true

[dependencies]
num-traits.workspace = true
paste = "1.0"
serde.workspace = true
serde_json.workspace = true
thiserror.workspace = true
smallvec.workspace = true

# Using workspace dependency for stwo
stwo-prover.workspace = true
