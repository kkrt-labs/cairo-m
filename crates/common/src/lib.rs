pub mod instruction;
pub mod opcode;
pub mod program;

pub use instruction::Instruction;
pub use opcode::Opcode;
pub use program::{Program, ProgramMetadata};
