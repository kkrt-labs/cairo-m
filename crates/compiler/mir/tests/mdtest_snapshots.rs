//! MIR snapshot tests for mdtest cases.
//! This file automatically generates MIR snapshots for all Cairo-M code in the mdtest directory,
//! providing comprehensive coverage of real-world examples through the MIR generation pipeline.

mod common;

use cairo_m_compiler_mir::{generate_mir, PrettyPrint};
use cairo_m_test_utils::{mdtest::MdTestRunner, mdtest_path};
use common::{create_test_crate, TestDatabase};

#[test]
fn test_mdtest_mir_snapshots() {
    use insta::{assert_snapshot, glob, with_settings};

    // Use glob! to discover all markdown files in mdtest directory
    glob!(mdtest_path().to_str().unwrap(), "**/*.md", |path| {
        let db = TestDatabase::default();

        let runner = MdTestRunner::new("MIR", |source, name| {
            let crate_id = create_test_crate(&db, source, name, "mdtest");

            match generate_mir(&db, crate_id) {
                Ok(module) => Ok(module.pretty_print(0)),
                Err(diagnostics) => Err(format!(
                    "MIR generation failed with diagnostics:\n{:#?}",
                    diagnostics
                )),
            }
        });

        let snapshots = runner.run_file(path);

        for snapshot in snapshots {
            with_settings!({
                description => format!("MIR snapshot for mdtest: {}", snapshot.name).as_str(),
                omit_expression => true,
                snapshot_suffix => snapshot.suffix,
                prepend_module_to_snapshot => false,
            }, {
                assert_snapshot!(snapshot.content);
            });
        }
    });
}
