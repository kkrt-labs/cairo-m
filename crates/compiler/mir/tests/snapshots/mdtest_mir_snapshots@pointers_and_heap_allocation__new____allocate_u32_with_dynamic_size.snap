---
source: crates/compiler/mir/tests/mdtest_snapshots.rs
assertion_line: 52
description: "MIR snapshot for mdtest: Pointers and Heap Allocation (new) - Allocate u32 with dynamic size"
input_file: mdtest/03-types/03-pointers-and-heap.md
---
Source:
fn alloc_u32() -> u32 {
    let p: u32* = new u32[3];
    // Single element write/read
    p[0] = 42u32;
    return p[0];
}
============================================================
Generated MIR:
module {
  // Function 0
  fn alloc_u32 {
    entry: 0

    0:
      %0 = 3 * 2
      %1 = heapalloccells %0
      %2 = load %1[0]
      store 42 -> %1[0]
      %3 = load %1[0]
      return %3

  }

}
